go/vt/wrangler/reparent.go

```diff
@@ -876,117 +883,101 @@ func (wr *Wrangler) emergencyReparentShardLocked(ctx context.Context, ev *events
+	if err != nil {
+		return err
 	}
-	wg.Wait()
 
 	// Check we still have the topology lock.
 	if err := topo.CheckShardLocked(ctx, keyspace, shard); err != nil {
 		return fmt.Errorf("lost topology lock, aborting: %v", err)
 	}
 
-	// Verify masterElect is alive and has the most advanced position
-	masterElectStatus, ok := statusMap[masterElectTabletAliasStr]
-	if !ok {
-		return fmt.Errorf("couldn't get master elect %v replication position", topoproto.TabletAliasString(masterElectTabletAlias))
-	}
-	masterElectStrPos := masterElectStatus.After.Position
-	masterElectPos, err := mysql.DecodePosition(masterElectStrPos)
+	validCandidates, err := wr.findValidReparentCandidates(statusMap, masterStatusMap)
 	if err != nil {
-		return fmt.Errorf("cannot decode master elect position %v: %v", masterElectStrPos, err)
+		return err
 	}
-	for alias, status := range statusMap {
-		if alias == masterElectTabletAliasStr {
-			continue
+	if len(validCandidates) == 0 {
+		return fmt.Errorf("no valid candidates for emergency reparent")
+	}
+
+	validCandidatesSet := sets.NewString(validCandidates...)
+	if masterElectTabletAlias != nil {
+		if !validCandidatesSet.Has(newMasterTabletAliasStr) {
+			return fmt.Errorf("master elect is either not fully caught up, or has errant GTIDs")
 		}
-		posStr := status.After.Position
-		pos, err := mysql.DecodePosition(posStr)
-		if err != nil {
-			return fmt.Errorf("cannot decode replica %v position %v: %v", alias, posStr, err)
+		validCandidatesSet = sets.NewString(newMasterTabletAliasStr)
+	}
+
+	// We run a race among our valid candidates.
+	// The first candidate (tablet) to succeed at applying its relay logs is the winner.
+	aliasChan := make(chan string)
+	rec := &concurrency.AllErrorRecorder{}
+	groupCtx, groupCancel := context.WithTimeout(ctx, waitReplicasTimeout)
+	defer groupCancel()
+	for candidate := range validCandidatesSet {
+		go func(alias string) {
+			var resultAlias string
+			defer func() { aliasChan <- resultAlias }()
+
+			err = wr.WaitForRelayLogsToApply(groupCtx, tabletMap[alias], statusMap[alias])
+			if err == nil {
+				resultAlias = alias
+			} else {
+				rec.RecordError(err)
+			}
+		}(candidate)
+	}
+
+	resultCounter := 0
+	var winningTabletAlias string
+	for alias := range aliasChan {
+		resultCounter++
+		if alias != "" && winningTabletAlias == "" {
+			winningTabletAlias = alias
+			// We only need the first tablet that succeeds, so we can cancel the other tablets
+			// and move along.
+			groupCancel()
 		}
-		if !masterElectPos.AtLeast(pos) {
-			return fmt.Errorf("tablet %v is more advanced than master elect tablet %v: %v > %v", alias, masterElectTabletAliasStr, posStr, masterElectStrPos)
+		if resultCounter == validCandidatesSet.Len() {
+			break
 		}
 	}
+	if winningTabletAlias == "" {
+		return fmt.Errorf("could not find a valid candidate for new master that applied its relay logs under the provided wait_replicas_timeout: %v", rec.Error())
```

Comment:
```suggestion
		return fmt.Errorf("could not find a valid candidate for new master that applied its relay logs within the provided wait_replicas_timeout: %v", rec.Error())
```

Reply:
I'm not picky about this suggestion, feel free to improve on it.

Reply:
Done.